sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -1476,9 +1476,11 @@ case class StringLocate(substr: Expression, str: Expression, start: Expression)
   override def third: Expression = start
   override def nullable: Boolean = substr.nullable || str.nullable
   override def dataType: DataType = IntegerType
-  override def inputTypes: Seq[DataType] = Seq(StringType, StringType, IntegerType)
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(StringTypeAnyCollation, StringTypeAnyCollation, IntegerType)
 
   override def eval(input: InternalRow): Any = {
+    val collationId = first.dataType.asInstanceOf[StringType].collationId
```

Comment:
consider moving this out and making it:
`final lazy val collationId: Int = first.dataType.asInstanceOf[StringType].collationId`